src/ast_editor.ml

```diff
@@ -436,24 +436,42 @@ module Command = struct
   let _reveal_ast_node =
     let handler (instance : Extension_instance.t) ~textEditor ~edit:_ ~args:_ =
       let document = TextEditor.document textEditor in
-      let webview_opt =
-        let ast_editor_state = Extension_instance.ast_editor_state instance in
-        Ast_editor_state.find_webview_by_doc ast_editor_state
-          (TextDocument.uri document)
-      in
-      let offset =
-        let selection = Vscode.TextEditor.selection textEditor in
-        let position = Vscode.Selection.start selection in
-        TextDocument.offsetAt document ~position
+      let (_ : unit Promise.t) =
+        let open Promise.Syntax in
+        let+ webview =
+          let ast_editor_state = Extension_instance.ast_editor_state instance in
+          match
+            Ast_editor_state.find_webview_by_doc ast_editor_state
+              (TextDocument.uri document)
+          with
+          | Some _ as r -> Promise.return r
+          | None ->
+            let* (_ : Ojs.t option) =
+              Commands.executeCommand
+                ~command:Extension_consts.Commands.open_ast_explorer_to_the_side
+                ~args:[]
+            in
+            (* without this small timeout, sometimes the command only opens an
+               AST explorer but doesn't reveal the node. *)
+            let+ () = Node.setTimeout' ~wait_ms:200 in
```

Comment:
While it's understandable why you've done it this way, I don't think we can use this. What's going to guarantee that the node will be revealed in 200ms for big documents, slower laptops, etc?

We need to think of a different of a different approach.  I'm looking at this `send_msg` function:

```
let send_msg t value ~(webview : WebView.t) =
  let msg = Ojs.empty_obj () in
  Ojs.set_prop_ascii msg "type" (Ojs.string_to_js t);
  Ojs.set_prop_ascii msg "value" value;
  let (_ : bool Promise.t) = WebView.postMessage webview msg in
  ()
```

perhaps we wait for the promise it returns?

Reply:
If we want to be on the safe side, I think we need to refactor `vscode.commands.registerCommand`'s `callback` to return `Ojs.t` rather than `unit` as it is the case currently. Then we register "open ast explorer" command not as a text editor command but an ordinary command, then we'll be able to wait for the command to complete before revealing a node.

Reply:
references: 
- https://code.visualstudio.com/api/references/vscode-api#commands
- https://github.com/microsoft/vscode/issues/16814

Reply:
Sounds good, but why do we need to go through the command api at all? It's untyped, error handling is not clear, and is meant to be user facing. Let's do things directly when possible.

Reply:
You're absolutely right. Pushed a fixed version. Seems to work. Tested on `lsp/types.ml` a couple of times